lib/etas/cluster_update.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+require "services"
+require "cluster_overlap"
+
+module ETAS
+  # Update DB's overlap records for every ht_item in a given cluster
+  class ClusterUpdate
+    attr_accessor :cluster
+
+    def initialize(overlap_table, cluster)
+      @overlap_table = overlap_table
+      @cluster = cluster
+    end
+
+    def deletes
+      @deletes ||= (existing_overlaps - new_overlaps)
+    end
+
+    def adds
+      @adds ||= (new_overlaps - existing_overlaps)
+    end
+
+    def new_overlaps
+      if @new_overlaps
+        return @new_overlaps
+      else
+        @new_overlaps = ClusterOverlap.new(cluster, nil).each.to_a.map(&:to_hash)
+      end
+
+      @new_overlaps ||= []
+    end
+
+    # Overlaps for ht_items currently in the cluster
+    # We will not find overlaps for items that have moved to another cluster,
+    # those will be deleted when the other cluster adds their new overlaps.
+    def existing_overlaps
```

Comment:
This seems a little bit clunky; I'll see if I can propose an alternative.

Reply:
This seems like it works:

```ruby
    def existing_overlaps
      @existing_overlaps ||= cluster.ht_items.map do |ht_item|
        @overlap_table.filter(volume_id: ht_item.item_id).map(&:to_hash)
      end.flatten
    end
```

If you wanted, you could extract the block body as a separate function, e.g.

```ruby
    def existing_overlaps
      @existing_overlaps ||= cluster.ht_items.map { |h| existing_item_overlap(h) }.flatten
    end

    private
   
    def existing_item_overlap(ht_item)
      @overlap_table.filter(volume_id: ht_item.item_id).map(&:to_hash)
    end
```

Reply:
I had something similar that was leaving me with nil and nil related errors which is how I ended up with my verbose clunker. Confirmed, yours works.